### DISCUSSION TITLE
Anybody know the variant storage size...?

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#1171] [Q&A]
From the VBA Docs (https://learn.microsoft.com/en-us/office/vba/language/reference/user-interface-help/data-type-summary):
<html>
<body>
<div class="has-inner-focus" style="box-sizing: inherit; outline-color: inherit; color: rgb(23, 23, 23); font-family: &quot;Segoe UI&quot;, SegoeUI, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Data type | Storage size | Range
-- | -- | --
Object | 4 bytes (32-bit) / 8 Bytes (64-bit) | Any Object reference
Variant (with numbers) | 16 bytes | Any numeric value up to the range of a Double
Variant (with characters) | 22 bytes + string length (24 bytes on 64-bit systems) | Same range as for variable-length String

</div><p style="box-sizing: inherit; outline-color: inherit; margin: 1rem 0px 0px; padding: 0px; overflow-wrap: break-word; color: rgb(23, 23, 23); font-family: &quot;Segoe UI&quot;, SegoeUI, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">A<span> </span><strong style="box-sizing: inherit; outline-color: inherit; font-weight: 600;">Variant</strong><span> </span>containing an array requires 12 bytes more than the array alone.</p></div>
</body>
</html>

So what does the memory usage for a Variant(Object) look like? Is it just the variant overhead + 4-byte object pointer? (and what is that overhead?) ...or something more?
Also, CVErr() is supposed to return a Variant(Error) - I presume that would match the Variant(Object) memory usage?

**Comment by bclothier** [comment]
The documentation is kind of confusing. I also think it's wrong for the Variant+String on 64-bit.

The [VARIANT structure](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/ns-oaidl-variant) is a fixed-size structure, 16-byte on 32-bit and 24 bytes on 64-bit system. The structure is a nested union with fixed size for the header section that is required for all supported data types EXCEPT decimal. The data section may contain unused bytes but it will never exceed more than 16-bytes or 24 bytes on 32-bit or 64-bit systems, respectively. Decimal is special because it occupies 16 bytes of memory BUT has a constant value in the first 2 bytes which corresponds to the `vt` discriminator indicating it's a decimal data type so it's really 14 bytes of useful data.  Of course, you can't align 14 bytes, so it's 16-bytes. 

But regardless of which data type you stuff in a variant, you are consuming 16 bytes or 24 bytes for the variant itself. If the data type is blittable, then no additional memory is needed. Strings and arrays may or may not be blittable but requires some pointer chasing. In the case of strings, it uses BSTR structure, so 6 bytes overhead for a zero length stringer is required. 4 bytes for the length, and 2 bytes for the mandatory null character terminator that is beyond the "allocated" region of the BSTR structure.  

What you are seeing in the linked documentation is trying to count the memory used both by the VARIANT structure and the referenced data whenever it's not a certain data type. Strings and Objects aren't "inside" a VARIANT structure and therefore the size will vary. But that's fairly a useless piece of information because you normally don't blit objects and strings require some handling which depend on the API you are using with the strings. 

`CVErr()` will use the `scode` union member so it's 4-bytes integer stored inside a 16 or 24 bytes structure.

Years ago I wrote a dumb C program to print out the offset and size consumed just to illustrate how the offset of certain members change between 32-bit and 64-bit systems. All fixed-size union members will have the same size while all union members that are a pointer will change between 4 to 8.

Reminder: Remember memory is read from left to right, so if a union member is on offset 4 but occupies 4 bytes of memory, then you need to read byte 4, 5, 6, and 7 to get that variable. Also remember it's zero-based so the first member is at offset zero.

32-bit:
```
The size of VARIANT on x32 is: 16
 and vt is at: 0
 and wReserved1 is at: 2, 2
 and wReserved2 is at: 4, 2
 and wReserved3 is at: 6, 2
 and llVal is at: 8, 8
 and lVal is at: 8, 4
 and bVal is at: 8, 1
 and iVal is at: 8, 2
 and fltVal is at: 8, 4
 and dblVal is at: 8, 8
 and boolVal is at: 8, 2
 and __OBSOLETE__VARIANT_BOOL is at: 8, 2
 and scode is at: 8, 4
 and cyVal is at: 8, 8
 and date is at: 8, 8
 and bstrVal is at: 8
 and punkVal is at: 8
 and pdispVal is at: 8
 and parray is at: 8
 and piVal is at: 8
 and plVal is at: 8
 and pllVal is at: 8
 and pfltVal is at: 8
 and pdblVal is at: 8
 and pboolVal is at: 8
 and __OBSOLETE__VARIANT_PBOOL is at: 8
 and pscode is at: 8
 and pcyVal is at: 8
 and pdate is at: 8
 and pbstrVal is at: 8
 and ppunkVal is at: 8
 and ppdispVal is at: 8
 and pparray is at: 8
 and pvarVal is at: 8
 and byref is at: 8
 and cVal is at: 8
 and uiVal is at: 8
 and ulVal is at: 8
 and ullVal is at: 8
 and intVal is at: 8
 and uintVal is at: 8
 and pdecVal is at: 8
 and pcVal is at: 8
 and puiVal is at: 8
 and pulVal is at: 8
 and pullVal is at: 8
 and pintVal is at: 8
 and puintVal is at: 8
 and pvRecord is at: 8
 and pRecInfo is at: 12
 and decVal is at: 0, 16
```

64-bit:
```
The size of VARIANT on x64 is: 24
 and vt is at: 0
 and wReserved1 is at: 2, 2
 and wReserved2 is at: 4, 2
 and wReserved3 is at: 6, 2
 and llVal is at: 8, 8
 and lVal is at: 8, 4
 and bVal is at: 8, 1
 and iVal is at: 8, 2
 and fltVal is at: 8, 4
 and dblVal is at: 8, 8
 and boolVal is at: 8, 2
 and __OBSOLETE__VARIANT_BOOL is at: 8, 2
 and scode is at: 8, 4
 and cyVal is at: 8, 8
 and date is at: 8, 8
 and bstrVal is at: 8
 and punkVal is at: 8
 and pdispVal is at: 8
 and parray is at: 8
 and piVal is at: 8
 and plVal is at: 8
 and pllVal is at: 8
 and pfltVal is at: 8
 and pdblVal is at: 8
 and pboolVal is at: 8
 and __OBSOLETE__VARIANT_PBOOL is at: 8
 and pscode is at: 8
 and pcyVal is at: 8
 and pdate is at: 8
 and pbstrVal is at: 8
 and ppunkVal is at: 8
 and ppdispVal is at: 8
 and pparray is at: 8
 and pvarVal is at: 8
 and byref is at: 8
 and cVal is at: 8
 and uiVal is at: 8
 and ulVal is at: 8
 and ullVal is at: 8
 and intVal is at: 8
 and uintVal is at: 8
 and pdecVal is at: 8
 and pcVal is at: 8
 and puiVal is at: 8
 and pulVal is at: 8
 and pullVal is at: 8
 and pintVal is at: 8
 and puintVal is at: 8
 and pvRecord is at: 8
 and pRecInfo is at: 16
 and decVal is at: 0, 16
```

**Reply by mburns08109** [reply]
Ok, I think I'm belatedly realizing that I've incorrectly conflated 2 different questions.
1) what is the storage size of these various data types  (i.e. for ETL/data interchange purposes) - the VISIBLE data storage requirements?
2) When used in a UDT, how much space does this take up? (a question for which, there is likely not just a single, simple answer due to the way differing elements in a UDT may combine and/or the way compilers may "optimize" things).
So perhaps I need to narrow my original question to the first question, omitting, for now, the second question's answers.
